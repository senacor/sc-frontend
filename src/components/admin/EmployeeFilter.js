import React, { useState } from 'react';
import { injectIntl } from 'react-intl';
import FilterList from '@material-ui/icons/FilterList';
import AddIcon from '@material-ui/icons/Add';
import Button from '@material-ui/core/Button';
import {
  Avatar,
  IconButton,
  List,
  ListItem,
  ListItemText,
  Popover,
  withStyles
} from '@material-ui/core';
import InputAdornment from '@material-ui/core/InputAdornment';
import Icon from '@material-ui/core/Icon';
import TextField from '@material-ui/core/TextField';
import { addMaintenanceTeamMember } from '../../calls/admin';

const styles = theme => ({
  box: {
    display: 'flex',
    padding: 2 * theme.spacing.unit,
    flexDirection: 'column'
  },
  leftIcon: {
    marginRight: theme.spacing.unit
  },
  pointer: {
    cursor: 'pointer'
  },
  spacing: {
    margin: 2 * theme.spacing.unit
  }
});

export const EmployeeFilter = ({
  data,
  setSelectedEmployee,
  maintenance,
  errorContext,
  intl,
  classes
}) => {
  const [anchorEl, setAnchorEl] = useState(null);
  const [value, setValue] = useState('');

  const handleOpen = event => {
    setAnchorEl(event.currentTarget);
    event.stopPropagation();
  };

  const handleClose = event => {
    setAnchorEl(null);
    event.stopPropagation();
  };

  const handleDelete = () => {
    setValue('');
    setSelectedEmployee(null);
  };

  const handleChange = event => {
    setValue(event.target.value);
  };

  const selectEmployee = (employee, event) => {
    setSelectedEmployee(employee);
    if (maintenance) {
      addMaintenanceTeamMember(employee.id, errorContext);
    }
    handleClose(event);
  };

  // TODO: value stays in the popup after closing

  return (
    <div>
      {maintenance ? (
        <Button
          color="primary"
          className={classes.spacing}
          onClick={handleOpen}
        >
          <AddIcon className={classes.leftIcon} />
          {intl.formatMessage({
            id: 'maintenance.add'
          })}
        </Button>
      ) : (
        <IconButton onClick={handleOpen}>
          <FilterList />
        </IconButton>
      )}
      <Popover
        open={Boolean(anchorEl)}
        anchorEl={anchorEl}
        onClose={handleClose}
        onClick={event => event.stopPropagation()}
      >
        <div className={classes.box}>
          <TextField
            id="outlined-adornment-filter"
            variant="outlined"
            label={intl.formatMessage({
              id: 'employeefilter.name'
            })}
            value={value}
            onChange={handleChange}
            onClick={event => {
              event.stopPropagation();
            }}
            InputProps={{
              endAdornment: (
                <InputAdornment position="end">
                  <IconButton onClick={handleDelete}>
                    <Icon id="adornmentIcon">clear</Icon>
                  </IconButton>
                </InputAdornment>
              ),
              name: 'employeeSearchValue'
            }}
          />
          {value && (
            <List>
              {data.map(employee => {
                return (
                  (employee.firstName
                    .toLowerCase()
                    .startsWith(value.toLowerCase()) ||
                    employee.lastName
                      .toLowerCase()
                      .startsWith(value.toLowerCase())) && (
                    <ListItem className={classes.pointer}>
                      <Avatar>
                        {employee.firstName.charAt(0)}
                        {employee.lastName.charAt(0)}
                      </Avatar>
                      <ListItemText
                        primary={`${employee.firstName} ${employee.lastName}`}
                        onClick={event => selectEmployee(employee, event)}
                      />
                    </ListItem>
                  )
                );
              })}
            </List>
          )}
        </div>
      </Popover>
    </div>
  );
};

export default injectIntl(withStyles(styles)(EmployeeFilter));
