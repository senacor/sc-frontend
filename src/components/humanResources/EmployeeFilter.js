import React, { Component } from 'react';
import * as actions from '../../actions';
import EmployeeSearch from '../employeeSearch/EmployeeSearch';
import { getFilter } from '../../reducers/selector';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

class EmployeeFilter extends Component {
  constructor(props) {
    super(props);

    let filter;
    let emptyFilter = {
      [this.props.filterBy]: { searchString: '', values: '' }
    };
    if (!this.props.filter || !this.props.filter[this.props.filterGroup]) {
      filter = emptyFilter;
    } else {
      filter = Object.assign(
        {},
        emptyFilter,
        this.props.filter[this.props.filterGroup]
      );
    }

    this.state = {
      employeeName: filter[this.props.filterBy].values,
      filter: filter
    };
  }

  selectEmployee = employee => {
    if (employee) {
      let filter = this.state.filter;
      filter[this.props.filterBy] = {
        searchString: `${this.props.filterBy}=${employee.id}`,
        values: `${employee.firstName} ${employee.lastName}`
      };
      let payload = {
        filterGroup: this.props.filterGroup,
        filter: filter
      };
      this.props.fetchFilteredPrsForHumanResource(payload);
    }
  };

  render() {
    return (
      <EmployeeSearch
        employeeSearchValue={this.state.employeeName}
        selectEmployee={this.selectEmployee}
      />
    );
  }
}

EmployeeFilter.propTypes = {
  filterGroup: PropTypes.string.isRequired,
  filterBy: PropTypes.string.isRequired
};

export default connect(
  state => ({
    filter: getFilter(state)
  }),
  {
    fetchFilteredPrsForHumanResource: actions.fetchFilteredPrsForHumanResource
  }
)(EmployeeFilter);
