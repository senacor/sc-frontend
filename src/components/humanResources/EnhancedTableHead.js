import React from 'react';
import TableHead from '@material-ui/core/TableHead/TableHead';
import TableRow from '@material-ui/core/TableRow/TableRow';
import TableCell from '@material-ui/core/TableCell/TableCell';
import TableSortLabel from '@material-ui/core/TableSortLabel/TableSortLabel';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core';

const styles = {
  cell: {
    textAlign: 'left'
  }
};

const EnhancedTableHead = props => {
  const createSortHandler = index => event => {
    props.onRequestSort(event, index);
  };

  return (
    <TableHead>
      <TableRow>
        {props.columnDefinition.map((column, index) => {
          return (
            <TableCell
              key={index}
              padding={'none'}
              sortDirection={props.orderBy === index ? props.order : false}
              variant={'head'}
              numeric={true}
              className={props.classes.cell}
            >
              <TableSortLabel
                active={props.orderBy === index}
                direction={props.order}
                onClick={createSortHandler(index)}
              >
                {column.filter ? column.filter : null}
                {column.label}
              </TableSortLabel>
            </TableCell>
          );
        }, this)}
      </TableRow>
    </TableHead>
  );
};

EnhancedTableHead.propTypes = {
  onRequestSort: PropTypes.func.isRequired,
  order: PropTypes.string.isRequired,
  orderBy: PropTypes.number.isRequired,
  columnDefinition: PropTypes.array.isRequired
};

const StyledComponent = withStyles(styles)(EnhancedTableHead);
export default StyledComponent;
