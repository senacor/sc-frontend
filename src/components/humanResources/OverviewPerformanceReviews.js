import React, { Component } from 'react';
import PerformanceReviewsTable from './PerformanceReviewsTable';
import { connect } from 'react-redux';
import {
  getAllPrsForTable,
  getFilter,
  getFilterPossibilities,
  isLoadingAction
} from '../../reducers/selector';
import * as actions from '../../actions';
import withLoading from '../hoc/Loading';
import FILTER_GROUPS from './filterGroups';
import { getUserroles } from '../../reducers/selector';
import { isHr } from '../../helper/checkRole';
import PerformanceReviewsTableService from './PerformanceReviewsTableService';
import { LoadingEvents } from '../../helper/loadingEvents';

export class OverviewPerformanceReviews extends Component {
  getColumnDefinitions = () => {
    const prTableService = new PerformanceReviewsTableService(
      FILTER_GROUPS.HR,
      this.props.filterPossibilities
    );

    return [
      prTableService.employee(),
      prTableService.deadline(),
      prTableService.occasion(),
      prTableService.projectCst(),
      prTableService.competence(),
      prTableService.level(),
      prTableService.supervisor(),
      prTableService.reviewer(),
      prTableService.result(),
      prTableService.employeePreparation(),
      prTableService.reviewerPreparation(),
      prTableService.meeting(),
      prTableService.finalState(),
      prTableService.hrDone()
    ];
  };

  componentDidUpdate(prevProps) {
    if (this.props.filter !== prevProps.filter) {
      this.props.fetchFilteredPrsForHumanResource(this.props.filter);
    }
  }

  render() {
    if (!this.props.filterPossibilities.levels) {
      return null;
    }

    let columns = this.getColumnDefinitions();
    let isHrMember = isHr(this.props.userroles);
    return (
      <PerformanceReviewsTable
        columnDefinition={columns}
        orderBy={1}
        data={this.props.data}
        filter={this.props.filter}
        isHr={isHrMember}
      />
    );
  }
}

export default connect(
  state => ({
    data: getAllPrsForTable(state),
    filter: getFilter(FILTER_GROUPS.HR)(state),
    userroles: getUserroles(state),
    isLoading: isLoadingAction(state, [
      LoadingEvents.FILTER_POSSIBILITIES,
      LoadingEvents.FETCH_OWN_PRS
    ]),
    filterPossibilities: getFilterPossibilities(state)
  }),
  {
    fetchFilteredPrsForHumanResource: actions.fetchFilteredPrsForHumanResource,
    getFilterPossibilities: actions.getFilterPossibilities
  }
)(
  withLoading(props => {
    props.getFilterPossibilities();
    props.fetchFilteredPrsForHumanResource(props.filter);
  })(OverviewPerformanceReviews)
);
