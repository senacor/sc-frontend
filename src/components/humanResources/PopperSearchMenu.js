import React, { useState } from 'react';
import Popover from '@material-ui/core/Popover';
import Icon from '@material-ui/core/Icon/Icon';
import IconButton from '@material-ui/core/IconButton/IconButton';
import { getSubFilter } from '../../reducers/selector';
import { connect } from 'react-redux';
import { withStyles } from '@material-ui/core';

const styles = theme => ({
  iconFilterSet: {
    color: theme.palette.primary['500'],
    margin: '0px'
  },
  iconFilterUnset: {
    color: '#dddddd',
    margin: '0px'
  }
});

export const PopperSearchMenu = ({
  classes,
  subfilter,
  children,
  filterBy,
  filterGroup
}) => {
  const [anchorEl, setAnchorEl] = useState(null);

  const handleClick = event => {
    setAnchorEl(event.currentTarget);
    event.stopPropagation();
  };

  const handleClose = event => {
    setAnchorEl(null);
  };

  const getIcon = subfilter => {
    return subfilter === '' ? (
      <Icon className={classes.iconFilterUnset}>filter_list</Icon>
    ) : (
      <Icon className={classes.iconFilterSet}>filter_list</Icon>
    );
  };

  const open = Boolean(anchorEl);
  return (
    <div>
      <IconButton onClick={handleClick}>{getIcon(subfilter)}</IconButton>
      <Popover
        id="simple-popper"
        open={open}
        anchorEl={anchorEl}
        onClose={handleClose}
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'center'
        }}
        transformOrigin={{
          vertical: 'top',
          horizontal: 'center'
        }}
        onClick={event => {
          event.stopPropagation();
        }}
      >
        <div>
          {React.cloneElement(children, {
            closeFilter: handleClose,
            filterBy: filterBy,
            filterGroup: filterGroup
          })}
        </div>
      </Popover>
    </div>
  );
};

export const StyledComponent = withStyles(styles)(PopperSearchMenu);
export default connect(
  (state, props) => ({
    subfilter: getSubFilter(props.filterGroup, props.filterBy)(state)
  }),
  {}
)(StyledComponent);
