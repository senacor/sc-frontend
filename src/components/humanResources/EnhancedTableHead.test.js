import React from 'react';
import EnhancedTableHead from './EnhancedTableHead';
import { createShallow } from '@material-ui/core/test-utils';
import HR_ELEMENTS from './hrElements';
import { Link } from 'react-router-dom';
import PopperSearchMenu from './PopperSearchMenu';
import EmployeeFilter from './EmployeeFilter';
import { getDisplayName } from './OverviewPerformanceReviews';
import FILTER_GROUPS from './filterGroups';

const rows = [
  {
    key: HR_ELEMENTS.EMPLOYEE,
    numeric: false,
    disablePadding: false,
    label: 'Mitarbeiter',
    mapper: variable => getDisplayName(variable),
    show: entry => {
      return (
        <Link to={`/prDetail/${entry.prId}`}>
          {getDisplayName(entry[HR_ELEMENTS.EMPLOYEE])}
        </Link>
      );
    },
    filter: (
      <PopperSearchMenu>
        <EmployeeFilter
          filterGroup={FILTER_GROUPS.HR}
          filterBy={HR_ELEMENTS.EMPLOYEE}
        />
      </PopperSearchMenu>
    )
  }
];

describe('EnhancedTableHead component', () => {
  let shallow = createShallow({ dive: true });

  it('should match snapshot', () => {
    const mockCallBack = jest.fn();
    let component = shallow(
      <EnhancedTableHead
        onRequestSort={mockCallBack}
        order={'asc'}
        orderBy={0}
        columnDefinition={rows}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('should have a clickable TableSortLabel', () => {
    const mockCallBack = jest.fn();

    const component = shallow(
      <EnhancedTableHead
        order={'asc'}
        orderBy={0}
        onRequestSort={mockCallBack}
        columnDefinition={rows}
      />
    );

    component.find('WithStyles(TableSortLabel)').simulate('click');
    expect(mockCallBack.mock.calls.length).toEqual(1);
  });
});
