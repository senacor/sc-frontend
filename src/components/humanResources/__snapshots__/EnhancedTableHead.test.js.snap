// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnhancedTableHead component should match snapshot 1`] = `
<WithStyles(TableHead)>
  <WithStyles(TableRow)>
    <WithStyles(TableCell)
      key="0"
      numeric={false}
      padding="none"
      sortDirection="asc"
    >
      <WithStyles(List)>
        <WithStyles(ListItem)
          className="EnhancedTableHead-tableCell-1"
          style={
            Object {
              "display": "flex",
              "justifyContent": "center",
            }
          }
        >
          <Connect(PopperSearchMenu)>
            <Connect(EmployeeFilter)
              filterBy="employee"
              filterGroup="hr"
            />
          </Connect(PopperSearchMenu)>
          <WithStyles(TableSortLabel)
            active={true}
            direction="asc"
            onClick={[Function]}
          >
            Mitarbeiter
          </WithStyles(TableSortLabel)>
        </WithStyles(ListItem)>
      </WithStyles(List)>
    </WithStyles(TableCell)>
  </WithStyles(TableRow)>
</WithStyles(TableHead)>
`;
